### PR TITLE
feat: runbooks v2

### PR DESCRIPTION
**Pull request by EmanuelJr**
## 📝 Description

This PR implements comprehensive runbook execution capabilities with support for multiple git repositories, caching, connection-based access control. The changes enable users to execute runbooks against specific connections while respecting role-based access rules.

Key additions include:
- Multi-repository runbook configuration with branch support
- Connection-level access control based on user groups and runbook rules
- Normalized Git URL handling for SSH and HTTPS repositories

## 🚀 Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🎨 Style/UI update
- [x] ♻️ Code refactor
- [ ] ⚡ Performance improvement
- [x] ✅ Test update
- [ ] 🔧 Build configuration change
- [ ] 🧹 Chore

### Core Runbook Models & Configuration
- Added `Branch` field to `Config` struct for git branch specification
- Implemented `NewConfigV2()` function with enhanced SSH and basic auth support
- Added `GetNormalizedGitURL()` method for consistent git URL handling (SSH and HTTPS)

### Git Integration
- Enhanced `CloneRepositoryInMemory()` to support custom branch selection with fallback to `master`/`main`
- Added `BuildRepositoryFromCommit()` function to build repositories from commit objects
- Improved branch resolution with dynamic fallback strategy

### Runbook Cache System
- Implemented nested `sync.Map` caching structure: `map[orgId]map[gitUrl]*runbookCache`
- Added `setRunbookCache()`, `GetRunbookCache()`, and `deleteRunbookCache()` functions
- Cache TTL support with 5-minute default expiration
- Selective cache invalidation by organization and git URL

### Access Control & Authorization
- Implemented `getRunbookConnections()` helper to filter connections based on runbook rules, connections and user groups
- Implemented admin-only bypass for access control

### Test Coverage
- Added comprehensive test suite in `helpers_test.go` covering:
  - Admin user scenarios
  - User group matching
  - Connection filtering with various rule combinations
  - Edge cases with empty configurations

### Test Configuration:
- **OS**: macOS

### Tests performed:
- [x] Unit tests pass
- [ ] Integration tests pass
- [x] Manual testing completed

### Test Cases Added:
- No runbook rules defined scenarios
- Admin user bypasses access control
- User group and connection matching
- Multiple matching rules with connection aggregation
- Edge cases with empty configurations and runbooks

## ✅ Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

## 📄 Additional Notes

---

### CODE REVIEW

File: common/runbooks/config.go [hunk 1]
```diff
@@ -25,6 +27,80 @@ type Config struct {
...
+	gitURL, err := base64.StdEncoding.DecodeString(envVars["GIT_URL"])
+	if err != nil {
+		return nil, ErrEmptyConfiguration
+	}
+
+	hookCacheTTL, err := parseRunbookHookCacheTTLConfig(envVars)
+	if err != nil {
+		return nil, err
+	}
+	gitBranch := envVars["GIT_BRANCH"]
```
Comment (sandromello) [inline common/runbooks/config.go]: The plugin configuration uses base64-encoded values, with clients responsible for encoding them before sending. Having some configuration values encoded while others remain in raw format would create an inconsistent and cumbersome developer experience.

Using a Git commit SHA would be more reliable than specifying a branch. Commit SHAs provide immutable references that are easier to track, whereas branches can change over time. But not a blocker though.

File: common/runbooks/config.go [hunk 2]
```diff
@@ -100,6 +176,55 @@ func NewConfig(envVars map[string]string) (*Config, error) {
 
 var globalSSHKeyScanKnownHostsContent string
 
+func (c *Config) GetNormalizedGitURL() string {
```
Comment (sandromello) [inline common/runbooks/config.go:162]: I think it's worth having unit tests here to see if this logic matches with your intention of normalizing repositories.

File: common/runbooks/config.go [hunk 3]
```diff
@@ -100,6 +176,55 @@ func NewConfig(envVars map[string]string) (*Config, error) {
...
+	}
+
+	host := parsed.Hostname()
+	path := strings.TrimSuffix(parsed.Path, ".git")
+	path = strings.Trim(path, "/")
+
+	return fmt.Sprintf("%s/%s", host, path)
+}
+
+func (c *Config) GetRepositoryName() string {
```
Comment (sandromello) [inline common/runbooks/config.go]: I think it's worth having unit tests here to see if this logic matches with your intention of parsing repository names.

File: common/runbooks/config.go [hunk 4]
```diff
@@ -133,7 +258,7 @@ func (c *Config) loadKnownHosts() (err error) {
 	// fallback loading from the plugin configuration
 	knownHostsContent, err := base64.StdEncoding.DecodeString(c.sshKnownHostsEnc)
 	if err != nil {
-		return fmt.Errorf("failed decoding GIT_SSH_KNOWN_HOSTS")
+		return fmt.Errorf("failed decoding SSH_KNOWN_HOSTS")
```
Comment (sandromello) [inline common/runbooks/config.go:219]: It was `GIT_SSH_KNOWN_HOSTS` because it's the name of the configuration. It's important to warn about the proper configuration name to avoid confusion when returning errors.

Any reason to change to `SSH_KNOWN_HOSTS`?

File: gateway/api/runbooks/helpers.go [hunk 5]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
 
 	"github.com/go-git/go-git/v5/plumbing/object"
 	"github.com/hoophq/hoop/common/runbooks"
 	"github.com/hoophq/hoop/gateway/api/openapi"
 	"github.com/hoophq/hoop/gateway/models"
+	"github.com/hoophq/hoop/gateway/storagev2/types"
 )
 
-const maxTemplateSize = 1000000 // 1MB
+const maxTemplateSize = 1_000_000 // 1MB
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go:18]: Wow, I wasn't aware of that syntax in GO. 🤯

File: gateway/api/runbooks/helpers.go [hunk 6]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+}
+
+var runbooksCache sync.Map // sync.Map[orgId]map[gitUrl]*runbookCache
+
+func setRunbookCache(orgId, gitUrl string, commit *object.Commit) {
+	// tenta obter o mapa interno
+	v, ok := runbooksCache.Load(orgId)
+	var inner map[string]*runbookCache
+	if ok {
+		inner = v.(map[string]*runbookCache)
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go]: This logic could panic. Make sure to cast it properly:

```go
inner, ok = v.(map[string]*runbookCache)
if !ok {
  // handle cast issue
}
```

File: gateway/api/runbooks/helpers.go [hunk 7]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+	} else {
+		inner = make(map[string]*runbookCache)
+		runbooksCache.Store(orgId, inner)
+	}
+
+	// atualiza o mapa interno
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go]: remove or change portuguese comments

File: gateway/api/runbooks/helpers.go [hunk 8]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+
+type runbookCache struct {
+	commit   *object.Commit
+	cachedAt time.Time
+}
+
+var runbooksCache sync.Map // sync.Map[orgId]map[gitUrl]*runbookCache
+
+func setRunbookCache(orgId, gitUrl string, commit *object.Commit) {
+	// tenta obter o mapa interno
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go]: remove or change portuguese comments

File: gateway/api/runbooks/helpers.go [hunk 9]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+	// atualiza o mapa interno
+	inner[gitUrl] = &runbookCache{
+		commit:   commit,
+		cachedAt: time.Now(),
+	}
+}
+
+func GetRunbookCache(orgId, gitUrl string) (*object.Commit, bool) {
+	if v, ok := runbooksCache.Load(orgId); ok {
+		inner := v.(map[string]*runbookCache)
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go]: It will panic if you not cast it properly.

File: gateway/api/runbooks/helpers.go [hunk 10]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+}
+
+func deleteRunbookCache(orgId string, gitUrl string) {
+	if gitUrl == "" {
+		runbooksCache.Delete(orgId)
+		return
+	}
+
+	if v, ok := runbooksCache.Load(orgId); ok {
+		inner := v.(map[string]*runbookCache)
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go]: It will panic if you not cast it properly.

File: gateway/api/runbooks/helpers.go [hunk 11]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+
+	return slices.ContainsFunc(a, func(x T) bool {
+		return slices.Contains(b, x)
+	})
+}
+
+func getRunbookConnections(runbookRules []models.RunbookRules, connectionList []string, runbookRepository, runbookName string, userGroups []string) []string {
+	// If no connections or rules are defined, return a list of empty connections
+	if len(connectionList) == 0 || len(runbookRules) == 0 {
+		return connectionList
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go:111]: A nil slice is functionally equivalent to an empty slice in Go, so you can safely return nil and check if the result is empty using len().
```go
items := getRunbookConnections(...)
if len(items) == 0 {
    // This works whether items is nil or an empty slice
}
```

File: gateway/api/runbooks/helpers.go [hunk 12]
```diff
@@ -2,15 +2,204 @@ package apirunbooks
...
+		for _, conn := range rule.Connections {
+			connectionsMap[conn] = struct{}{}
+		}
+	}
+
+	// Transform map keys to slice
+	connections := make([]string, 0, len(connectionsMap))
+	for conn := range connectionsMap {
+		connections = append(connections, conn)
+	}
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go:152]: Heads up: You'll lose order here, just a warning if it could be a problem

File: gateway/api/runbooks/rules.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+// @Router       /runbooks/rules [get]
+func ListRunbookRules(c *gin.Context) {
+	ctx := storagev2.ParseContext(c)
+
+	rules, err := models.GetRunbookRules(models.DB, ctx.GetOrgID(), 0, 0)
+	if err != nil {
+		log.Infof("failed fetching runbook rules, err=%v", err)
+		c.JSON(http.StatusUnprocessableEntity, gin.H{"message": "failed fetching runbook rules"})
+		return
+	}
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go:35]: I think it's worth to differentiate between common error from internal errors here.
If you're not expecting any errors, you should treat as 500 not 422

File: gateway/api/runbooks/rules.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+)
+
+// ListRunbookRules
+//
+// @Summary      List Runbook Rules
+// @Description  List all Runbook Rules
+// @Tags         Runbooks
+// @Produce      json
+// @Success      200  {object}  []openapi.RunbookRule
+// @Failure      404,422,500  {object} openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/rules.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	ctx := storagev2.ParseContext(c)
+
+	ruleID := c.Param("id")
+
+	rule, err := models.GetRunbookRuleByID(models.DB, ctx.GetOrgID(), ruleID)
+	if err != nil {
+		log.Infof("failed fetching runbook rule, err=%v", err)
+		c.JSON(http.StatusNotFound, gin.H{"message": "runbook rule not found"})
+		return
+	}
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go:70]: It's not always a not found error. Make sure to differentiate between not found and internal errors

File: gateway/api/runbooks/rules.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+// GetRunbookRule
+//
+// @Summary      Get Runbook Rule
+// @Description  Get a single Runbook Rule by ID
+// @Tags         Runbooks
+// @Produce      json
+// @Param        id  path      string  true  "Runbook Rule ID"
+// @Success      200  {object}  openapi.RunbookRule
+// @Failure      404,422,500  {object} openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/rules.go [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+// @Router       /runbooks/rules/{id} [get]
+func GetRunbookRule(c *gin.Context) {
+	ctx := storagev2.ParseContext(c)
+
+	ruleID := c.Param("id")
+
+	rule, err := models.GetRunbookRuleByID(models.DB, ctx.GetOrgID(), ruleID)
+	if err != nil {
+		log.Infof("failed fetching runbook rule, err=%v", err)
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: Do not print Info messages to not found errors, only internal errors or something that you really want to know that happened. Use debugf if you want to keep this line.

File: gateway/api/runbooks/rules.go [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+// CreateRunbookRule
+//
+// @Summary      Create Runbook Rule
+// @Description  Create a new Runbook Rule
+// @Tags         Runbooks
+// @Accept       json
+// @Produce      json
+// @Param        rule  body     openapi.RunbookRule  true  "Runbook Rule"
+// @Success      201  {object}  openapi.RunbookRule
+// @Failure      400,422,500  {object} openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/rules.go [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+// CreateRunbookRule
+//
+// @Summary      Create Runbook Rule
+// @Description  Create a new Runbook Rule
+// @Tags         Runbooks
+// @Accept       json
+// @Produce      json
+// @Param        rule  body     openapi.RunbookRule  true  "Runbook Rule"
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: I think the proper name for this is `request` not `rule`.

File: gateway/api/runbooks/rules.go [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+		CreatedAt:   rule.CreatedAt,
+		UpdatedAt:   rule.UpdatedAt,
+	}
+}
+
+func buildRunbookRuleModelFromRequest(reqRule *openapi.RunbookRule) models.RunbookRules {
+	id := reqRule.ID
+	if id == "" {
+		id = uuid.NewString()
+	}
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: You shouldn't allow the client to choose the id for this resource when creating it.

File: gateway/api/runbooks/rules.go [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+	var reqRule openapi.RunbookRule
+	if err := c.ShouldBindJSON(&reqRule); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"message": "invalid request body"})
+		return
+	}
+
+	rule := buildRunbookRuleModelFromRequest(&reqRule)
+	rule.OrgID = ctx.GetOrgID()
+
+	if err := models.UpsertRunbookRule(models.DB, &rule); err != nil {
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go:100]: You must validate if the resource exists in order to reliable use this function. The POST should only be able to create the resource not to updated it in case it exists.

File: gateway/api/runbooks/rules.go [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+// @Failure      400,404,422,500  {object} openapi.HTTPError
+// @Router       /runbooks/rules/{id} [put]
+func UpdateRunbookRule(c *gin.Context) {
+	ctx := storagev2.ParseContext(c)
+
+	ruleID := c.Param("id")
+
+	var reqRule openapi.RunbookRule
+	if err := c.ShouldBindJSON(&reqRule); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"message": "invalid request body"})
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: Just return the error to the client, it will help us to identify issues in the client without having to inspect gateway logs.

File: gateway/api/runbooks/rules.go [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+		return
+	}
+
+	rule := buildRunbookRuleModelFromRequest(&reqRule)
+	rule.ID = ruleID
+	rule.OrgID = ctx.GetOrgID()
+
+	if err := models.UpsertRunbookRule(models.DB, &rule); err != nil {
+		log.Errorf("failed updating runbook rule, reason=%v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "failed updating runbook rule"})
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: Just return the error to the client, it will help us to identify issues in the client without having to inspect gateway logs.

File: gateway/api/runbooks/rules.go [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+		c.JSON(http.StatusBadRequest, gin.H{"message": "invalid request body"})
+		return
+	}
+
+	rule := buildRunbookRuleModelFromRequest(&reqRule)
+	rule.OrgID = ctx.GetOrgID()
+
+	if err := models.UpsertRunbookRule(models.DB, &rule); err != nil {
+		log.Errorf("failed creating runbook rule, reason=%v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "failed creating runbook rule"})
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: Just return the error to the client, it will help us to identify issues in the client without having to inspect gateway logs.

File: gateway/api/runbooks/rules.go [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+// @Param        rule  body     openapi.RunbookRule  true  "Runbook Rule"
+// @Success      201  {object}  openapi.RunbookRule
+// @Failure      400,422,500  {object} openapi.HTTPError
+// @Router       /runbooks/rules [post]
+func CreateRunbookRule(c *gin.Context) {
+	ctx := storagev2.ParseContext(c)
+
+	var reqRule openapi.RunbookRule
+	if err := c.ShouldBindJSON(&reqRule); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"message": "invalid request body"})
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: Just return the error to the client, it will help us to identify issues in the client without having to inspect gateway logs.

File: gateway/api/runbooks/rules.go [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+// @Failure      404,422,500  {object} openapi.HTTPError
+// @Router       /runbooks/rules/{id} [delete]
+func DeleteRunbookRule(c *gin.Context) {
+	ctx := storagev2.ParseContext(c)
+
+	ruleID := c.Param("id")
+
+	if err := models.DeleteRunbookRule(models.DB, ctx.GetOrgID(), ruleID); err != nil {
+		log.Errorf("failed deleting runbook rule, reason=%v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "failed deleting runbook rule"})
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: Just return the error to the client, it will help us to identify issues in the client without having to inspect gateway logs.

File: gateway/api/runbooks/rules.go [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+//
+// @Summary      Update Runbook Rule
+// @Description  Update an existing Runbook Rule by ID
+// @Tags         Runbooks
+// @Accept       json
+// @Produce      json
+// @Param        id    path      string              true  "Runbook Rule ID"
+// @Param        rule  body      openapi.RunbookRule  true  "Runbook Rule"
+// @Success      200  {object}   openapi.RunbookRule
+// @Failure      400,404,422,500  {object} openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/rules.go [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+
+// DeleteRunbookRule
+//
+// @Summary      Delete Runbook Rule
+// @Description  Delete a Runbook Rule by ID
+// @Tags         Runbooks
+// @Produce      json
+// @Param        id  path      string  true  "Runbook Rule ID"
+// @Success      204  {object}  nil
+// @Failure      404,422,500  {object} openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/rules.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/runbooksV2.go [hunk 29]
```diff
@@ -0,0 +1,434 @@
...
+)
+
+// ListRunbooks
+//
+//	@Summary		List Runbooks
+//	@Description	List all Runbooks
+//	@Tags			Runbooks
+//	@Produce		json
+//	@Success		200			{object}	openapi.RunbookList
+//	@Failure		404,422,500	{object}	openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/runbooksV2.go [hunk 30]
```diff
@@ -0,0 +1,434 @@
...
+
+	runbookRules, err := models.GetRunbookRules(models.DB, ctx.OrgID, 0, 0)
+	if err != nil {
+		log.Infof("failed fetching runbook rules, err=%v", err)
+		c.JSON(http.StatusUnprocessableEntity, gin.H{"message": "failed fetching runbook rules"})
+		return
+	}
+
+	urlQuery := c.Request.URL.Query()
+	connection := urlQuery.Get("connection")
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go:61]: Add the query string to the openapi spec documentation

File: gateway/api/runbooks/runbooksV2.go [hunk 31]
```diff
@@ -0,0 +1,434 @@
...
+		connectionNames = []string{connection}
+		removeEmptyConnectionsList = true
+	} else {
+		connectionNames, err = models.ListConnectionsName(models.DB, ctx.GetOrgID())
+		if err != nil {
+			log.Infof("failed fetching connection names, err=%v", err)
+			c.JSON(http.StatusUnprocessableEntity, gin.H{"message": "failed fetching connection names"})
+			return
+		}
+	}
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go]: There's no need the else branch:

```go
connectionNames := []string{connection}
if connection == "" {
  connectionNames, err = models.ListConnectionsName(models.DB, ctx.GetOrgID())
  (...)
}
```

File: gateway/api/runbooks/runbooksV2.go [hunk 32]
```diff
@@ -0,0 +1,434 @@
...
+
+	runbooks, err := models.GetRunbookConfigurationByOrgID(models.DB, ctx.GetOrgID())
+	if err != nil {
+		if errors.Is(err, models.ErrNotFound) {
+			c.JSON(http.StatusNotFound, gin.H{"message": "runbook configuration not found"})
+			return
+		}
+
+		log.Infof("failed fetching runbook configuration, err=%v", err)
+		c.JSON(http.StatusUnprocessableEntity, gin.H{"message": "failed fetching runbook configuration"})
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go]: If it's not a not found error, it's an internal error. Make sure to return the message to the client as well

File: gateway/api/runbooks/runbooksV2.go [hunk 33]
```diff
@@ -0,0 +1,434 @@
...
+
+// GetRunbookConfiguration
+//
+//	@Summary		Get Runbook Configuration
+//	@Description	Get Runbook Configuration
+//	@Tags			Runbooks
+//	@Accept			json
+//	@Produce		json
+//	@Success		200			{object}	openapi.RunbookConfiguration
+//	@Failure		400,404,422,500	{object}	openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go]: These errors must match what you're returning from the function

File: gateway/api/runbooks/runbooksV2.go [hunk 34]
```diff
@@ -0,0 +1,434 @@
...
+// UpdateRunbookConfiguration
+//
+//	@Summary		Update Runbook Configuration
+//	@Description	Update Runbook Configuration
+//	@Tags			Runbooks
+//	@Accept			json
+//	@Produce		json
+//	@Param			runbook	body		openapi.RunbookConfigurationRequest	true	"Runbook Configuration"
+//	@Success		200			{object}	openapi.RunbookConfiguration
+//	@Failure		400,404,422,500	{object}	openapi.HTTPError
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go:159]: These errors must match what you're returning from the function

File: gateway/api/runbooks/runbooksV2.go [hunk 35]
```diff
@@ -0,0 +1,434 @@
...
+		if _, isset := repositoryConfigs[config.GetNormalizedGitURL()]; isset {
+			c.JSON(http.StatusUnprocessableEntity, gin.H{"message": "duplicate git repository URLs are not allowed"})
+			return
+		}
+
+		repositoryConfigs[config.GetNormalizedGitURL()] = mapConfig
+	}
+
+	runbooks := models.Runbooks{
+		ID:                uuid.NewString(),
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go]: If you're creating a new uuid, it will not update the record on subsequent calls to this endpoint.

File: gateway/api/runbooks/runbooksV2.go [hunk 36]
```diff
@@ -0,0 +1,434 @@
...
+		OrgID:             ctx.GetOrgID(),
+		RepositoryConfigs: repositoryConfigs,
+	}
+
+	err := models.UpsertRunbookConfiguration(models.DB, &runbooks)
+	if err != nil {
+		log.Errorf("failed upserting runbook, reason=%v", err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "failed upserting runbook"})
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go]: Make sure to propagate the error to the client

File: common/runbooks/config.go [hunk 37]
```diff
@@ -25,6 +27,80 @@ type Config struct {
...
+	hookCacheTTL, err := parseRunbookHookCacheTTLConfig(envVars)
+	if err != nil {
+		return nil, err
+	}
+	gitBranch := envVars["GIT_BRANCH"]
+	gitUserEnc := envVars["GIT_USER"]
+	gitPasswordEnc := envVars["GIT_PASSWORD"]
+	sshKeyEnc := envVars["SSH_KEY"]
+	sshUserEnc := envVars["SSH_USER"]
+	sshKeyPassEnc := envVars["SSH_KEY_PASS"]
```
Comment (sandromello) [inline common/runbooks/config.go]: If the goal is not reusing the plugin configuration you could use raw values for all configurations. The plugin configuration was a generic structure, that why we used base64 encoded values. 

If your structure will not rely on values that are not serialized in JSON, just use raw values

File: gateway/api/runbooks/runbooksV2.go [hunk 38]
```diff
@@ -0,0 +1,434 @@
...
+
+	repoConfig, ok := configs.RepositoryConfigs[req.Repository]
+	if !ok {
+		c.JSON(http.StatusBadRequest, gin.H{"message": fmt.Sprintf("runbook repository config %v not found", req.Repository)})
+		return
+	}
+
+	config, err := runbooks.NewConfigV2(repoConfig)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go:329]: Make sure to log internal errors as well

File: gateway/api/runbooks/runbooksV2.go [hunk 39]
```diff
@@ -0,0 +1,434 @@
...
+		return
+	}
+
+	commit, ok := GetRunbookCache(ctx.GetOrgID(), config.GetNormalizedGitURL())
+	if !ok {
+		var err error
+		commit, err = runbooks.CloneRepositoryInMemory(config)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
```
Comment (sandromello) [inline gateway/api/runbooks/runbooksV2.go:339]: Make sure to log internal errors as well

File: gateway/models/runbooks.go [hunk 40]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/hoophq/hoop/gateway/storagev2/types"
+	"github.com/lib/pq"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+type Runbooks struct {
+	ID                string                       `gorm:"column:id"`
+	OrgID             string                       `gorm:"column:org_id"`
+	RepositoryConfigs map[string]map[string]string `gorm:"column:repository_configs;serializer:json"`
```
Comment (sandromello) [inline gateway/models/runbooks.go]: I recommend always using well-defined types in Go that serialize properly to JSON. Initially, I started with maps, but later learned how to implement this correctly using structured types. While proceeding with maps isn't necessarily problematic, using explicit types for models makes the code clearer and significantly improves API documentation.

Some examples:
- `./gateway/models/server_auth_config.go`
- `./gateway/models/datamasking.go`

File: rootfs/app/migrations/000048_add_runbooks_table.up.sql [hunk 41]
```diff
@@ -0,0 +1,27 @@
...
+  UNIQUE (org_id)
+);
+
+CREATE TABLE runbook_rules (
+  id UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+  org_id UUID NOT NULL REFERENCES orgs (id),
+  name TEXT NOT NULL,
+  description TEXT,
+  user_groups TEXT[] NOT NULL,
+  connections TEXT[] NOT NULL,
```
Comment (sandromello) [inline rootfs/app/migrations/000048_add_runbooks_table.up.sql:21]: This could result in a very long list if the user has many connections. Have you considered this before creating this type of association?

For other features (guardrails, datamasking), we've been using associative tables to manage relationships. Don't you think that would be a better approach here as well? With the current structure, it will be cumbersome to perform modifications in this structure.

File: gateway/models/runbooks.go [hunk 42]
```diff
@@ -0,0 +1,163 @@
...
+			"updated_at":  time.Now().UTC(),
+		}),
+	}).Create(rule)
+
+	return tx.Error
+}
+
+func DeleteRunbookRule(db *gorm.DB, orgID, ruleID string) error {
+	tx := db.Table("private.runbook_rules").Where("id = ? AND org_id = ?", ruleID, orgID).Delete(&RunbookRules{})
+	return tx.Error
```
Comment (sandromello) [inline gateway/models/runbooks.go]:
```go
return db.Table("private.runbook_rules").
  Where("id = ? AND org_id = ?", ruleID, orgID).
  Delete(&RunbookRules{}).
  Error
```

File: gateway/transport/extensions/runbookhook.go [hunk 43]
```diff
@@ -93,40 +96,138 @@ func getRunbookHookFiles(ctx Context) (*runbookWrapperFiles, error) {
...
+	runbooksConfig, err := models.GetRunbookConfigurationByOrgID(models.DB, ctx.OrgID)
+	if err != nil {
+		if err == models.ErrNotFound {
+			return hooks, nil
+		}
+		return nil, err
+	}
+
+	if cached, ok := hookStoreV2.Load(ctx.OrgID); ok {
+		cachedHooks := cached.([]*runbookWrapperFiles)
```
Comment (sandromello) [inline gateway/transport/extensions/runbookhook.go]: You must type cast this properly or it will panic

File: gateway/transport/extensions/runbookhook.go [hunk 44]
```diff
@@ -93,40 +96,138 @@ func getRunbookHookFiles(ctx Context) (*runbookWrapperFiles, error) {
...
+}
+
 func processEventOpenSessionHook(ctx Context, pkt *proto.Packet) {
 	hook, err := getRunbookHookFiles(ctx)
 	if err != nil {
 		log.With("sid", ctx.SID).Warnf("session-open: failed fetching runbook hook, reason=%v", err)
 		return
 	}
-	if hook == nil || hook.sessionOpen == nil {
+	hooksV2, err := getRunbookHookFilesV2(ctx)
```
Comment (sandromello) [inline gateway/transport/extensions/runbookhook.go:195]: If the user has an old configuration, how this would be handled? It's a breaking change?

**Review by sandromello — CHANGES_REQUESTED** [review]

File: gateway/api/runbooks/helpers.go [hunk 45]
```diff
@@ -2,15 +2,217 @@ package apirunbooks
...
+			inner = make(map[string]*runbookCache)
+		}
+	} else {
+		inner = make(map[string]*runbookCache)
+		runbooksCache.Store(orgId, inner)
+	}
+
+	inner[gitUrl] = &runbookCache{
+		commit:   commit,
+		cachedAt: time.Now(),
```
Comment (sandromello) [inline gateway/api/runbooks/helpers.go]: Always use time.Now().UTC()

File: gateway/api/runbooks/helpers.go (hunk 10 above)
Comment (sandromello) [reply gateway/api/runbooks/helpers.go]: It still not solved

**Review by sandromello — APPROVED** [review]
Just make sure to generate the openapi docs again.
